server/src/main/java/org/apache/druid/discovery/BaseNodeRoleWatcher.java

```diff
@@ -209,13 +221,59 @@ private void removeNode(DiscoveryDruidNode druidNode)
     }
   }
 
+  public void cacheInitializedTimedOut()
+  {
+    synchronized (lock) {
+      LOGGER.warn("Cache for node role [%s] could not be initialized before timeout.", nodeRole.getJsonName());
+      // No need to wait on CountDownLatch, because we are holding the lock under which it could only be
+      // counted down.
+      if (cacheInitialized.getCount() == 0) {
+        LOGGER.warn("Cache for node watcher of role[%s] is already initialized. ignoring timeout.", nodeRole.getJsonName());
+        return;
+      }
+
+      // It is important to take a snapshot here as list of nodes might change by the time listeners process
+      // the changes.
+      List<DiscoveryDruidNode> currNodes = Lists.newArrayList(nodes.values());
+      LOGGER.info(
+          "Node watcher of role [%s] is now initialized with %d nodes.",
+          nodeRole.getJsonName(),
+          currNodes.size());
+
+      for (DruidNodeDiscovery.Listener listener : nodeListeners) {
+        safeSchedule(
+            () -> {
+              listener.nodesAdded(currNodes);
```

Comment:
This logic is very similar to `cacheInitialized()`, it just calls the timed-out version instead. Just as a matter of code structure and maintainability, it would be good to extract shared logic to a private method like `cacheInitialized(boolean timedOut)`, and have both `cacheInitializedTimedOut()` and `cacheInitialized()` call that method.

Also `cacheInitializedTimedOut()` should be private (or package-private if you're calling it from unit tests), rather than public. It's not part of the API for callers.